runtime/memprof.c

```diff
@@ -451,48 +592,54 @@ void caml_memprof_track_young(tag_t tag, uintnat wosize, int from_caml)
   Caml_state->young_ptr += whsize;
   caml_memprof_renew_minor_sample();
 
-  /* Empty the queue to make sure callbacks are called in the right
-     order. */
-  caml_raise_if_exception(caml_memprof_handle_postponed_exn());
-
+  caml_memprof_suspended = 1;
   callstack = capture_callstack();
-  ephe = caml_raise_if_exception(do_callback_exn(tag, wosize, occurrences,
-                                                 callstack, Minor));
+  t_idx = new_tracked(occurrences, Make_header(wosize, tag, Caml_white),
+                      0, 1, Placeholder_value, callstack);
+  res = (t_idx == Invalid_index) ? run_any_callback_exn(&t_idx) : Val_unit;
+  caml_memprof_suspended = 0;
+  caml_memprof_check_action_pending();
```

Comment:
TODO: check whether this line may result in the `action_pending` flag being set at each minor-allocation sampling, which would be bad for performance.

Reply:
This line may actually set `pending_actions` because the `run_entry_callbacks_exn` just above did not update `trackst.callback`. Suggestion: have `run_entry_callbacks_exn` (or `handle_entry_callcbacks_exn`?) increment `threast.callback` at its entry point if it is called at this smallest index, even if no callback ends up being called. This should also simplify the logic in `caml_memprof_handle_postponed`.